bilge-impl/src/shared.rs

```diff
@@ -176,8 +115,99 @@ pub fn is_always_filled(ty: &Type) -> bool {
     ty.starts_with('u') || ty == "bool"
 }
 
+pub fn enum_fills_bitsize(bitsize: u8, variants_count: usize) -> bool {
+    let max_variants_count = 1u128 << bitsize;
+    variants_count as u128 == max_variants_count
+}
 
 #[inline]
 pub fn unreachable<T, U>(_: T) -> U {
     unreachable!("should have already been validated")
 }
+
+fn fallback_variant(data: &Data) -> Option<Variant> {
+    match data {
+        Data::Enum(enum_data) => {
+            let mut variants_with_fallback = enum_data
+                .variants
+                .iter()
+                .filter(|variant| variant.attrs.iter().any(is_fallback_attribute));
+
+            let variant = variants_with_fallback.next();
+
+            if variants_with_fallback.next().is_some() {
+                abort_call_site!("only one enum variant may be fallback"; help = "remove #[fallback] attributes until you only have one");
+            } else {
+                variant.cloned()
+            }
+        }
+        Data::Struct(struct_data) => {
+            let mut field_attrs = struct_data.fields.iter().flat_map(|field| &field.attrs);
+            
+            if field_attrs.any(is_fallback_attribute) {
+                abort_call_site!("the attribute `fallback` is only applicable to enums"; help = "remove all `#[fallback]` from this struct")
+            } else {
+                None
+            }
+        }
+        _ => unreachable(())
+    }
+}
+
+pub fn is_attribute(attr: &Attribute, name: &str) -> bool {
+    if let Meta::Path(path) = &attr.meta {
+        path.is_ident(name)
+    } else {
+        false
+    }
+}
+
+fn is_non_exhaustive_attribute(attr: &Attribute) -> bool {
+    is_attribute(attr, "non_exhaustive")
+}
+
+pub(crate) fn is_fallback_attribute(attr: &Attribute) -> bool {
+    is_attribute(attr, "fallback")
+}
+
+pub(crate) struct EnumVariantValueAssigner {
+    bitsize: u8,
+    next_expected_assignment: u128,
+}
+
+impl EnumVariantValueAssigner {
+    pub fn new(bitsize: u8) -> EnumVariantValueAssigner {
+        EnumVariantValueAssigner { bitsize, next_expected_assignment: 0 }
+    }
+    
+    fn max_value(&self) -> u128 {
+        2u128.saturating_pow(self.bitsize as u32) - 1
```

Comment:
Nice, that's the "mathematic" version of `max_variants_count`. Let's decide for one of these:
`max_variants_count = 1u128 << bitsize`, `val >= max`
`max_variants_count = 2u128^bitsize - 1`, `val > max`

bitsize = 2, val = 3
`1 << 2 == 4`, `3 >= 4 == false`
`2^2 -1 == 3`, `3 > 3 == false`

Reply:
it is faster, but `1u128 << bitsize` would panic for a `bitsize` of 128, while the `saturating_pow` version would not.

Reply:
oh! I didn't notice. so yeah, no reason not to use a bitshift.
are you okay with: `(1u128 << bitsize) - 1` and `val > max`?

Reply:
yes

Reply:
done. thanks.